zygrader/bobs_shake.py

```diff
@@ -0,0 +1,422 @@
+
+    "shaking" occurs in a series of steps, roughly:
+        - user selects start and end time to include
+        - user points Bob to the data from the help queue
+        - data is read from zygrader's log and the queue info
+        - errors encountered and corrections that need to be made
+          are presented to the user
+        - input data is assigned to each TA appropriately
+        - each TA's events are analyzed to calculate the results summary
+        - user selects an output file
+        - the summary statistics are written to the file
+    """
+    window = ui.get_window()
+    wait_popup = ui.layers.WaitPopup("Bob's Shake")
+
+    worker = _StatsWorker()
+    Step = namedtuple('Step', ['interactive', 'msg', 'func'])
+    steps = [
+        Step(True, "start time", lambda: worker.select_start_time()),
+        Step(True, "end time", lambda: worker.select_end_time()),
+        Step(True, "select helpqueue csv",
+             lambda: worker.select_help_queue_data_file()),
+        Step(False, "Read data from the log file",
+             lambda: worker.read_in_native_stats()),
+        Step(False, "Read data from help queue file",
+             lambda: worker.read_in_help_queue_stats()),
+        Step(True, "present queue data errors",
+             lambda: worker.present_queue_errors()),
+        Step(True, "validate ta names", lambda: worker.validate_queue_names()),
+        Step(False, "Assign events to individual tas",
+             lambda: worker.assign_events_to_tas()),
+        Step(False, "Analyze stats for each TA",
+             lambda: worker.analyze_tas_individually()),
+        Step(True, "select output file", lambda: worker.select_output_file()),
+        Step(False, "Write shaken stats to file",
+             lambda: worker.write_stats_to_file())
+    ]
+
+    _WorkEvent.queue_errors = []
+
+    for step in steps:
+        if step.interactive:
+            if not step.func():
+                return
+        else:
+            wait_popup.set_message([step.msg])
+            wait_popup.set_wait_fn(step.func)
+            window.run_layer(wait_popup)
+            if wait_popup.canceled:
+                return
+
+
+class _WorkEvent:
+    queue_errors = []
+
+    def __init__(self, time_stamp, event_type, student_name, ta_name, is_begin,
+                 og_data, uniq_item):
+        self.time_stamp = time_stamp
+        self.event_type = event_type
+        self.student_name = student_name
+        self.ta_name = ta_name
+        self.is_begin = is_begin
+        self.og_data = og_data
+        self.uniq_item = uniq_item
+
+    @classmethod
+    def from_native_data(cls, row):
+        time_stamp_str = row[0]
+        # TODO: convert to datetime.datetime.fromisoformat
+        # once we are using python 3.7
+        time_stamp = datetime.datetime.strptime(time_stamp_str,
+                                                "%Y-%m-%dT%H:%M:%S.%f")
+
+        # The old lock format did not have an event_type field
```

Comment:
```suggestion
        # The old lock format did not have an event_type field
        # (all locks were for labs)
        is_old_format = len(row) == 5
        if is_old_format:
            row.insert(1, "LAB")

        event_type = row[1]
        student_name = row[2]
        assignment_name = row[3]
        ta_netid = row[4]
        
        ...
        
        lock_type = row[5]
```

I am not sure if keeping `row` intact matters, but this is more readable to me

Reply:
it doesn't matter if row stays intact, and I agree. I'll change this